src/GovUk.Education.ExploreEducationStatistics.Common.Tests/Cache/CacheAspectThreadManagementIntegrationTests.cs

```diff
@@ -0,0 +1,570 @@
+    
+    /// <summary>
+    /// This test is testing the flow of events is as we would expect to see it when the CacheAspect code is
+    /// non-blocking and in particular the scenario when an existing cache entry already exists.
+    ///
+    /// It fires off an HTTP request to our TestController which uses the BlobCache attribute on its
+    /// called method, and records the flow of events from there. It compares these with an expected order that would
+    /// be different had the CacheAspect introduced blocking code.
+    ///
+    /// Unfortunately we can't tell if there is further blocking code in one of the multiple continuations that will be
+    /// run as a part of this, but they would get picked up in the <see cref="NoThreadExhaustion"/> test very quickly
+    /// if they existed. 
+    /// </summary>
+    [Fact]
+    public async Task CachingControllerMethod_ExistingCachedItem()
+    {
+        // This is a list of events that we will capture in order to verify that the flow of code is showing what
+        // we expect to see, especially in terms of allowing threads to be freed up and not be blocked. 
+        var events = new List<Event>();
+        
+        var blobCacheService = new TestBlobCacheService(events);
+        var app = SetupApp(blobCacheService, events);
+
+        var client = app.CreateClient();
+
+        var expectedCacheKey = new TestCacheKey("myRequest");
+        var expectedControllerResult = new ControllerResponse("existingItem");
+
+        await blobCacheService.SetItemAsync(new TestCacheKey("myRequest"), new ControllerResponse("existingItem"));
+        events.Clear();
+        
+        var response = await client.GetAsync("/test/caching-method?key=myRequest");
+
+        response.AssertOk(expectedControllerResult);
+        
+        // A reasonable amount of time for code to reach the next step of the HTTP request processing workflow if
+        // no async delaying action appears between the two steps. 
+        var noDelayTimeMillis = AsyncOperationDelayMillis / 2;
+
+        // The first 4 items in this list of events are the most important orders of Events here. They show that the
+        // CacheAspect is non-blocking in this scenario, as we can see that:
+        //
+        // 4. We then see this evidence - the AsyncActionFilter.OnActionExecutionAsync() completes, with a
+        //    non-Completed Task as its result. This indicates that there is nothing in the CacheAspect code any more
+        //    that is tying up the Controller Request thread, which is now freed up to be used again.
+        // Previously, when the CacheAspect code was blocking the thread, we would see all of the work up to and
+        // including the setting of the cached item being done *within* the AsyncActionFilter.OnActionExecutionAsync()
+        // method, with the OnActionExecutionAsync method completing just prior to the AsyncResultFilter kicking in to
+        // return the HTTP response to the caller.
```

Comment:
Not sure what's gone on here, but have these notes been mangled somehow?

Reply:
Yup indeed, reordered now thanks